File: p11-platform/services/data-engine/connectors/claude_natural_connector.py
"""
Claude Natural Connector (Two-Phase GEO)
Phase 1: Get natural conversational response (like real Claude)
Phase 2: Analyze that response to extract GEO metrics
"""
import os
import logging
import re
from typing import Any, Dict, List, Optional, Tuple
import anthropic
from tenacity import retry, stop_after_attempt, wait_exponential

logger = logging.getLogger(__name__)


def extract_domain(url: str) -> str:
    try:
        from urllib.parse import urlparse
        parsed = urlparse(url)
        return (parsed.hostname or url).replace('www.', '', 1)
    except Exception:
        return url


def _block_value(block: Any, key: str, default: Any = None) -> Any:
    if isinstance(block, dict):
        return block.get(key, default)
    return getattr(block, key, default)


def extract_sources_from_claude_response(response: Any) -> List[Dict[str, str]]:
    """Pull citation URLs from Anthropic web_search content and text citations."""
    sources: List[Dict[str, str]] = []
    seen = set()

    def add_source(url: Optional[str], title: str = '', snippet: str = '') -> None:
        if not url or url in seen:
            return
        seen.add(url)
        sources.append({
            'title': title or '',
            'url': url,
            'domain': extract_domain(url),
            'snippet': snippet or '',
        })

    for block in _block_value(response, 'content', []) or []:
        for citation in _block_value(block, 'citations', []) or []:
            add_source(
                _block_value(citation, 'url'),
                _block_value(citation, 'title', '') or '',
                _block_value(citation, 'cited_text', '') or '',
            )

        if _block_value(block, 'type') != 'web_search_tool_result':
            continue
        content = _block_value(block, 'content', []) or []
        if isinstance(content, list):
            for item in content:
                add_source(
                    _block_value(item, 'url'),
                    _block_value(item, 'title', '') or '',
                    _block_value(item, 'snippet', '') or '',
                )

    return sources


class ClaudeNaturalConnector:
    """Two-phase natural mode connector for Claude."""
    
    def __init__(self):
        self.api_key = os.environ.get('ANTHROPIC_API_KEY')
        if not self.api_key:
            raise ValueError("ANTHROPIC_API_KEY not set")
        
        self.client = anthropic.Anthropic(api_key=self.api_key)
        self.model = os.environ.get('GEO_CLAUDE_MODEL', 'claude-sonnet-5')
        claude_search = os.environ.get('GEO_CLAUDE_WEB_SEARCH')
        if claude_search is not None:
            self.enable_web_search = claude_search.lower() == 'true'
        else:
            self.enable_web_search = os.environ.get('GEO_ENABLE_WEB_SEARCH', 'true').lower() == 'true'
        
        logger.info(f"[ClaudeNatural] Model: {self.model}, Web search: {self.enable_web_search}")
    
    @retry(stop=stop_after_attempt(3), wait=wait_exponential(multiplier=1, min=4, max=10))
    async def get_natural_response(self, query_text: str) -> Tuple[str, List[Dict], Dict]:
        """
        Phase 1: Get natural conversational response with web search.
        NO property context - simulates real user Claude experience.
        
        Returns:
            (response_text, search_sources, raw_response)
        """
        logger.info(f"[Claude-Natural] Phase 1: Getting natural response for: {query_text[:50]}...")
        
        system_prompt = 'You are a helpful assistant. Answer naturally in conversational prose. Do not output JSON. If unsure, say so plainly.'
        search_sources: List[Dict[str, str]] = []

        def collect_text(response: Any) -> str:
            return '\n'.join(
                _block_value(block, 'text', '') or ''
                for block in _block_value(response, 'content', []) or []
                if _block_value(block, 'type') == 'text'
            ).strip()

        if self.enable_web_search:
            try:
                logger.info("[Claude-Natural] Using Anthropic web_search_20250305")
                response = self.client.messages.create(
                    model=self.model,
                    max_tokens=2000,
                    system=system_prompt,
                    messages=[{"role": "user", "content": query_text}],
                    tools=[{
                        "type": "web_search_20250305",
                        "name": "web_search",
                        "max_uses": 5,
                    }],
                )
                search_sources = extract_sources_from_claude_response(response)
                content = collect_text(response)
                logger.info(
                    f"[Claude-Natural] Phase 1 complete: {len(content)} chars, {len(search_sources)} sources"
                )
                return (
                    content,
                    search_sources,
                    {
                        'response_id': response.id,
                        'model': self.model,
                        'usage': {
                            'input_tokens': response.usage.input_tokens,
                            'output_tokens': response.usage.output_tokens,
                        },
                        'stop_reason': response.stop_reason,
                        'used_web_search': True,
                    },
                )
            except Exception as error:
                logger.error(
                    "[Claude-Natural] Web search failed, falling back to no web search: %s",
                    error,
                    exc_info=True,
                )

        response = self.client.messages.create(
            model=self.model,
            max_tokens=2000,
            system=system_prompt,
            messages=[{"role": "user", "content": query_text}],
        )
        content = collect_text(response) or (response.content[0].text if response.content else '')
        logger.info(f"[Claude-Natural] Phase 1 complete: {len(content)} chars")
        return (
            content,
            [],
            {
                'response_id': response.id,
                'model': self.model,
                'usage': {
                    'input_tokens': response.usage.input_tokens,
                    'output_tokens': response.usage.output_tokens,
                },
                'stop_reason': response.stop_reason,
                'used_web_search': False,
            },
        )
    
    @retry(stop=stop_after_attempt(3), wait=wait_exponential(multiplier=1, min=4, max=10))
    async def analyze_response(self, context: Dict[str, Any]) -> Dict[str, Any]:
        """
        Phase 2: Analyze the natural response and extract GEO metrics.
        
        Args:
            context: Must include naturalResponse, brandName, queryText, brandDomains
            
        Returns:
            Structured analysis with answer_block and analysis metadata
        """
        natural_response = context['naturalResponse']
        brand_name = context['brandName']
        query_text = context['queryText']
        brand_domains = context.get('brandDomains', [])
        expected_city = context.get('expectedCity', '')
        expected_state = context.get('expectedState', '')
        
        logger.info(f"[Claude-Natural] Phase 2: Analyzing response for brand: {brand_name}")
        
        location_context = f"{expected_city}, {expected_state}" if expected_city and expected_state else "not specified"
        
        analysis_prompt = f"""You are analyzing an LLM's response to extract GEO visibility metrics.

Original Query: {query_text}

LLM Response to Analyze:
---
{natural_response}
---

Brand Information:
- Brand Name: {brand_name}
- Expected Location: {location_context}
- Brand Domains: {', '.join(brand_domains) if brand_domains else 'unknown'}
- Provider search sources: {', '.join(source.get('url') for source in context.get('searchSources') or [] if source.get('url')) or 'None provided.'}

Task: Extract structured data from the LLM response above. Return ONLY a JSON object with:

1. answer_block: The structured GEO data
   - ordered_entities: Array of apartment properties mentioned, in order
     Each entity: {{name, domain, rationale, position}}
   - citations: Array of URLs mentioned
     Each citation: {{url, domain, entity_ref}}
   - answer_summary: Brief summary of what the LLM said
   - notes.flags: Quality flags if applicable

2. analysis: Metadata about the extraction
   - brand_mentioned: boolean
   - brand_prominence: high/medium/low/none
   - extraction_confidence: 0.0-1.0
   - ordered_entities: Detailed extraction data

CRITICAL:
- Extract EVERY named property, community, builder, or listing brand. An empty ordered_entities array is only valid if the response names none of those.
- If {brand_name} is mentioned, it MUST appear in ordered_entities
- Position numbers start at 1 (the first mentioned property)
- Only include properties that were actually mentioned in the response
- If location doesn't match {location_context}, add flag "nap_mismatch"
- Do not flag possible_hallucination when {brand_name} is named in the response.
- You may include provider search sources in citations even when the prose has no URLs.

Output ONLY valid JSON, no markdown."""

        try:
            response = self.client.messages.create(
                model=self.model,
                max_tokens=4000,  # Increased to avoid truncation in detailed analysis
                system='You are a precise GEO extraction system. Output ONLY valid JSON without markdown or extra text.',
                messages=[
                    {"role": "user", "content": analysis_prompt}
                ]
            )
            
            content = response.content[0].text
            
            # Parse JSON
            import json
            try:
                parsed = json.loads(content)
            except json.JSONDecodeError:
                match = re.search(r'\{[\s\S]*\}', content)
                if match:
                    parsed = json.loads(match.group(0))
                else:
                    raise ValueError("Could not parse JSON from analysis")
            
            logger.info("[Claude-Natural] Phase 2 complete")
            
            # Ensure structure
            if 'answer_block' not in parsed:
                parsed = {'answer_block': parsed, 'analysis': {}}
            
            answer_block = parsed.get('answer_block', {})
            
            if 'ordered_entities' not in answer_block:
                answer_block['ordered_entities'] = []
            if 'citations' not in answer_block:
                answer_block['citations'] = []
            if 'answer_summary' not in answer_block:
                answer_block['answer_summary'] = natural_response[:200]
            if 'notes' not in answer_block:
                answer_block['notes'] = {'flags': []}
            
            return {
                'envelope': parsed,
                'raw': {
                    'response_id': response.id,
                    'model': self.model,
                    'usage': {
                        'input_tokens': response.usage.input_tokens,
                        'output_tokens': response.usage.output_tokens
                    }
                }
            }
            
        except Exception as e:
            logger.error(f"[Claude-Natural] Phase 2 error: {e}", exc_info=True)
            return {
                'envelope': {
                    'answer_block': {
                        'ordered_entities': [],
                        'citations': [],
                        'answer_summary': natural_response[:200],
                        'notes': {'flags': ['possible_hallucination']}
                    },
                    'analysis': {'error': str(e)}
                },
                'raw': {'error': str(e)}
            }
    
    async def invoke_natural_mode(self, context: Dict[str, Any]) -> Dict[str, Any]:
        """Complete two-phase natural mode execution."""
        # Phase 1
        natural_text, search_sources, phase1_raw = await self.get_natural_response(
            context['queryText']
        )
        
        # Phase 2
        analyzed = await self.analyze_response({
            'naturalResponse': natural_text,
            'brandName': context['brandName'],
            'queryText': context['queryText'],
            'brandDomains': context.get('brandDomains', []),
            'expectedCity': context.get('propertyLocation', {}).get('city'),
            'expectedState': context.get('propertyLocation', {}).get('state'),
            'searchSources': search_sources,
        })
        
        from connectors.entity_fallback import finalize_answer_block
        from connectors.evaluator import reconcile_citation_flags

        answer_block = analyzed['envelope']['answer_block']
        existing_citations = answer_block.get('citations', []) or []
        existing_urls = {citation.get('url') for citation in existing_citations if citation.get('url')}
        for source in search_sources:
            if source.get('url') and source['url'] not in existing_urls:
                existing_citations.append({
                    'url': source['url'],
                    'domain': source.get('domain') or extract_domain(source['url']),
                    'entity_ref': None,
                })
                existing_urls.add(source['url'])
        answer_block['citations'] = existing_citations
        notes = answer_block.get('notes') or {}
        notes['flags'] = reconcile_citation_flags(notes.get('flags') or [], len(existing_citations))
        answer_block['notes'] = notes
        answer_block = finalize_answer_block(
            answer_block,
            context,
            natural_text,
            analyzed['envelope'].get('analysis'),
        )
        logger.info(
            f"[Claude-Natural] Two-phase complete: {len(search_sources)} web sources, "
            f"{len(answer_block.get('citations') or [])} total citations, "
            f"{len(answer_block.get('ordered_entities') or [])} entities"
        )
        
        return {
            'answer': answer_block,
            'raw': {
                'audit_mode': 'natural',
                'phase1': phase1_raw,
                'phase2': analyzed['raw'],
                'natural_response': natural_text,
                'search_sources': search_sources,
                'analysis': analyzed['envelope'].get('analysis', {})
            }
        }

File: p11-platform/apps/web/utils/propertyaudit/claude-natural-connector.ts
/**
 * Claude Natural Connector for PropertyAudit (Two-phase GEO)
 * Phase 1: Natural conversational response (no property context)
 * Phase 2: Analyzer extracts structured GEO fields from the natural response
 * 
 * Uses Anthropic's native web_search_20250305 tool to capture actual LLM sources
 */

import Anthropic from '@anthropic-ai/sdk'
import {
  type NaturalConnector,
  type NaturalResponse,
  type NaturalAnalyzeContext,
  type NaturalAnalyzeResult,
  type WebSearchSource,
  NaturalExtractionEnvelopeSchema,
  getGeoConfig,
} from './types'

const ALLOWED_FLAGS = new Set([
  'no_sources',
  'possible_hallucination',
  'outdated_info',
  'nap_mismatch',
  'conflicting_prices'
] as const)

function coerceFlags(obj: unknown): unknown {
  if (!obj || typeof obj !== 'object') return obj
  const record = obj as Record<string, unknown>
  
  // Coerce answer_block.notes.flags
  const answerBlock = record.answer_block as Record<string, unknown> | undefined
  if (answerBlock?.notes && typeof answerBlock.notes === 'object') {
    const notes = answerBlock.notes as Record<string, unknown>
    if (Array.isArray(notes.flags)) {
      notes.flags = notes.flags.filter(
        (f): f is string => typeof f === 'string' && ALLOWED_FLAGS.has(f as any)
      )
    }
  }
  
  return record
}

function tryParseJson(content: string): unknown {
  // direct parse
  try {
    return JSON.parse(content)
  } catch {
    // continue
  }

  // Strip markdown code blocks
  let cleaned = content.trim()
  cleaned = cleaned.replace(/^```(?:json)?\s*\n?/i, '')
  cleaned = cleaned.replace(/\n?```\s*$/i, '')
  cleaned = cleaned.trim()

  try {
    return JSON.parse(cleaned)
  } catch {
    // continue
  }

  const firstBrace = cleaned.indexOf('{')
  const lastBrace = cleaned.lastIndexOf('}')
  if (firstBrace !== -1 && lastBrace !== -1 && lastBrace > firstBrace) {
    const jsonCandidate = cleaned.substring(firstBrace, lastBrace + 1)
    try {
      return JSON.parse(jsonCandidate)
    } catch {
      // continue
    }
  }

  const match = content.match(/\{[\s\S]*\}$/)
  if (match) {
    try {
      return JSON.parse(match[0])
    } catch {
      // ignore
    }
  }

  return null
}

function buildAnalyzerPrompt(ctx: NaturalAnalyzeContext): string {
  const expected =
    ctx.expectedCity && ctx.expectedState ? `${ctx.expectedCity}, ${ctx.expectedState}` : 'Unknown'

  return [
    `You are a GEO audit analyzer extracting structured data from a natural LLM response.`,
    ``,
    `IMPORTANT: Be objective. Extract what was ACTUALLY said, not what should have been said.`,
    `Do not invent citations or URLs.`,
    ``,
    `Original Query: "${ctx.queryText}"`,
    `Brand Being Tracked: "${ctx.brandName}"`,
    `Expected Location: ${expected}`,
    `Known brand domains (for inference only): ${ctx.brandDomains.join(', ') || '—'}`,
    `Known competitor domains (for inference only): ${ctx.competitors.join(', ') || '—'}`,
    `Provider search sources (API metadata; may not appear as URLs in the prose):`,
    !ctx.searchSources?.length
      ? 'None provided.'
      : ctx.searchSources
        .filter(source => source.url)
        .map((source, index) => `${index + 1}. ${source.title || source.domain || source.url} — ${source.url}`)
        .join('\n'),
    ``,
    `LLM's Natural Response to Analyze:`,
    `"""`,
    ctx.naturalResponse,
    `"""`,
    ``,
    `Return ONLY raw JSON with this exact schema:`,
    `{`,
    `  "answer_block": {`,
    `    "ordered_entities": [{"name":"...","domain":"...","rationale":"...","position":1}],`,
    `    "citations": [{"url":"...","domain":"...","entity_ref":"..."}],`,
    `    "answer_summary": "...",`,
    `    "notes": {"flags": []}`,
    `  },`,
    `  "analysis": {`,
    `    "ordered_entities": [{"name":"...","domain":"...","position":1,"prominence":"primary","mention_count":1,"first_mention_quote":"..."}],`,
    `    "citations": [{"url":"...","domain":"...","citation_type":"explicit"}],`,
    `    "brand_analysis": {"mentioned": true, "position": 1, "location_stated": "City, ST", "location_correct": true, "prominence": "primary"},`,
    `    "extraction_confidence": 85`,
    `  }`,
    `}`,
    ``,
    `Rules:`,
    `- Extract EVERY named property, community, builder, or listing brand. An empty ordered_entities array is only valid if the response names none of those.`,
    `- If the tracked brand is named, it MUST appear in answer_block.ordered_entities with a 1-based position.`,
    `- Do not flag possible_hallucination when the tracked brand is named in the response.`,
    `- answer_block.ordered_entities MUST be ordered by prominence (best-effort).`,
    `- Put the first_mention_quote into answer_block.ordered_entities[].rationale (include it verbatim).`,
    `- You may include provider search sources in citations even when the prose has no URLs.`,
    `- Do not invent extra URLs. no_sources is set later from the final citation list; do not flag it when provider sources were supplied.`,
  ].join('\n')
}

/**
 * Extract domain from URL
 */
function extractDomain(url: string): string {
  try {
    const urlObj = new URL(url)
    return urlObj.hostname.replace(/^www\./, '')
  } catch {
    return url
  }
}

/**
 * Extract source URLs from Claude's web search tool results
 */
function extractSourcesFromResponse(response: any): WebSearchSource[] {
  const sources: WebSearchSource[] = []
  const seenUrls = new Set<string>()

  // Look for web search results in the response content
  for (const item of response.content || []) {
    // Check for tool_result blocks that contain web search results
    if (item.type === 'tool_result' && item.content) {
      // Parse web search results from tool content
      try {
        const results = typeof item.content === 'string' 
          ? JSON.parse(item.content) 
          : item.content
        
        if (Array.isArray(results)) {
          for (const result of results) {
            if (result.url && !seenUrls.has(result.url)) {
              seenUrls.add(result.url)
              sources.push({
                title: result.title || '',
                url: result.url,
                domain: extractDomain(result.url),
                snippet: result.snippet || result.description || ''
              })
            }
          }
        }
      } catch {
        // Content wasn't JSON, try to extract URLs from text
      }
    }
    
    // Also check for citations in text blocks (newer Claude responses)
    if (item.type === 'text' && item.citations) {
      for (const citation of item.citations || []) {
        if (citation.url && !seenUrls.has(citation.url)) {
          seenUrls.add(citation.url)
          sources.push({
            title: citation.title || '',
            url: citation.url,
            domain: extractDomain(citation.url),
            snippet: citation.snippet || ''
          })
        }
      }
    }
  }

  return sources
}

export class ClaudeNaturalConnector implements NaturalConnector {
  surface = 'claude' as const

  async getNaturalResponse(query: string): Promise<NaturalResponse> {
    const config = getGeoConfig()
    const client = new Anthropic({ 
      apiKey: config.anthropicApiKey,
      timeout: 600000, // 10 minutes timeout for slow API responses
    })
    const enableWebSearch = process.env.GEO_ENABLE_WEB_SEARCH === 'true'

    const messages: Anthropic.MessageParam[] = [
      { role: 'user', content: query },
    ]

    // NOTE: We intentionally do NOT provide property/location context here.
    const systemPrompt = 'You are a helpful assistant. Answer naturally in conversational prose. Do not output JSON. If unsure, say so plainly.'

    if (enableWebSearch) {
      // Use Anthropic's native web_search_20250305 tool
      // This captures the actual sources Claude uses (like consumer Claude.ai)
      try {
        console.log('[claude-natural] Using native web search tool')
        
        const response = await client.messages.create({
          model: config.claudeModel,
          max_tokens: 2000,
          temperature: config.temperature,
          system: systemPrompt,
          messages,
          tools: [
            {
              type: 'web_search_20250305',
              name: 'web_search',
              max_uses: 5,
            } as any, // Type assertion needed for newer tool type
          ],
        })

        // Extract text from response
        const textBlocks = (response.content ?? []).filter(
          (b): b is Anthropic.TextBlock => b.type === 'text'
        )
        const text = textBlocks.map(b => b.text).join('\n')

        // Extract sources from the response
        const searchSources = extractSourcesFromResponse(response)
        
        // Also try to extract from any web_search_tool_result blocks
        for (const block of response.content || []) {
          if ((block as any).type === 'web_search_tool_result') {
            const results = (block as any).results || []
            for (const result of results) {
              if (result.url) {
                const exists = searchSources.some(s => s.url === result.url)
                if (!exists) {
                  searchSources.push({
                    title: result.title || '',
                    url: result.url,
                    domain: extractDomain(result.url),
                    snippet: result.snippet || ''
                  })
                }
              }
            }
          }
        }

        console.log(`[claude-natural] Extracted ${searchSources.length} sources from LLM response`)

        return {
          text,
          model: config.claudeModel,
          tokensUsed: (response.usage?.input_tokens ?? 0) + (response.usage?.output_tokens ?? 0),
          usedWebSearch: true,
          searchSources,
          rawResponse: response,
        }
      } catch (error) {
        console.error('[claude-natural] Web search tool error, falling back to no web search:', error)
        // Fall through to no web search
      }
    }

    // Fallback: No web search
    const response = await client.messages.create({
      model: config.claudeModel,
      max_tokens: 1400,
      temperature: config.temperature,
      system: systemPrompt,
      messages,
    })

    const textBlocks = (response.content ?? []).filter(
      (b): b is Anthropic.TextBlock => b.type === 'text'
    )
    const text = textBlocks.map(b => b.text).join('\n')

    return {
      text,
      model: config.claudeModel,
      tokensUsed: (response.usage?.input_tokens ?? 0) + (response.usage?.output_tokens ?? 0),
      usedWebSearch: false,
      searchSources: [],
      rawResponse: response,
    }
  }

  async analyzeResponse(context: NaturalAnalyzeContext): Promise<NaturalAnalyzeResult> {
    const config = getGeoConfig()
    const client = new Anthropic({ 
      apiKey: config.anthropicApiKey,
      timeout: 600000, // 10 minutes timeout for slow API responses
    })

    const prompt = buildAnalyzerPrompt(context)
    const response = await client.messages.create({
      model: config.claudeModel,
      max_tokens: 4000, // Increased to avoid truncation
      temperature: 0,
      system:
        'You are a precise GEO extraction system. Output ONLY valid JSON without markdown or extra text.',
      messages: [{ role: 'user', content: prompt }],
    })

    const raw = response
    const textBlocks = (response.content ?? []).filter(
      (b): b is Anthropic.TextBlock => b.type === 'text'
    )
    const contentText = textBlocks.map(b => b.text).join('\n')
    const rawJsonValue = tryParseJson(contentText)

    if (rawJsonValue === null) {
      console.error('[claude-natural] JSON parsing failed - not valid JSON. Raw:', contentText.slice(0, 800))
      throw new Error('Failed to parse natural extraction envelope: response was not valid JSON')
    }

    // Coerce invalid flags to prevent Zod validation failures
    const jsonValue = coerceFlags(rawJsonValue)

    const parsed = NaturalExtractionEnvelopeSchema.safeParse(jsonValue)
    if (!parsed.success) {
      console.error('[claude-natural] Zod validation failed. Raw:', contentText.slice(0, 800))
      console.error('[claude-natural] Zod errors:', JSON.stringify(parsed.error.issues, null, 2))
      const firstError = parsed.error.issues?.[0]?.message || parsed.error.message || 'Schema validation failed'
      throw new Error(`Failed to parse natural extraction envelope JSON: ${firstError}`)
    }

    return { envelope: parsed.data, raw }
  }
}










